python/prima.py

```diff
@@ -0,0 +1,341 @@
+try:
+    # import from current dir
+    import numpy as np
+
+    _path = os.path.abspath(os.path.dirname(__file__))
+    _handle = np.ctypeslib.load_library(_lib_name, _path)
+except Exception:
+    # else rely on loader env variables
+    _lib_loc = ctypes.util.find_library(_lib_name)
+    if _lib_loc is None:
+        raise ValueError(f"Cannot find {_lib_name} library")
+    _handle = ct.cdll.LoadLibrary(_lib_loc)
+
+# primac callback form
+_prima_function_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double)
+)
+_prima_function_con_callback = ct.CFUNCTYPE(
+    None, ct.POINTER(ct.c_double), ct.POINTER(ct.c_double),
+    ct.POINTER(ct.c_double)
+)
+
+# prima_get_rc_string returns char*
+_handle.prima_get_rc_string.restype = ct.c_char_p
+
+# pure Python function converters
+
+
+def _prima_pure_callback(func, n):
+    def _inner(cx, cf):
+        x = [cx[i] for i in range(n)]
+        f = func(x)
+        cf[0] = f
+
+    return _prima_function_callback(_inner)
+
+
+def _prima_pure_con_callback(func, n, m):
+    def _inner(cx, cf, ccon):
+        x = [cx[i] for i in range(n)]
+        f, con = func(x)
+        cf[0] = f
+        for i in range(m):
+            ccon[i] = con[i]
+
+    return _prima_function_con_callback(_inner)
+
+
+@dataclass
+class Result:
+    """Optimization result."""
+    x: list
+    success: bool
+    status: int
+    message: str
+    fun: float
+    nfev: int
+
+
+def bobyqa(f, n, x0, xl, xu, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
+    """
+    BOBYQA solver.
+
+    Parameters
+    ----------
+    f : callable
+        objective function, returns a float
+    n : int
+        input dimension
+    x0 : sequence of float
+        starting point
+    xl, xu : sequence of float
+        x bounds
+    rhobeg, rhoend : float
+        trust region radius
+    maxfun : int
+        max function evaluation
+    iprint : int
+        verbosity level
+
+    Returns
+    -------
+    res : Result
+        Result class
+    """
+    assert len(x0) == n
+    assert len(xl) == n
+    assert len(xu) == n
+    cn = ct.c_int(n)
+    cx = (ct.c_double * n)(*x0)
+    cf = ct.pointer(ct.c_double())
+    cxl = (ct.c_double * n)(*xl)
+    cxu = (ct.c_double * n)(*xu)
+    cnf = ct.pointer(ct.c_int())
+    crhobeg = ct.c_double(rhobeg)
+    crhoend = ct.c_double(rhoend)
+    cmaxfun = ct.c_int(maxfun)
+    ciprint = ct.c_int(iprint)
+    cfun = _prima_pure_callback(f, n)
+    status = _handle.prima_bobyqa(cfun, cn, cx, cf, cxl, cxu,
+            cnf, crhobeg, crhoend, cmaxfun, ciprint)
+    x = [cx[i] for i in range(n)]
+    success = (status == 0)
+    cstatus = ct.c_int(status)
+    message = _handle.prima_get_rc_string(cstatus).decode()
+    fun = cf[0]
+    nfev = cnf[0]
+    return Result(x, success, status, message, fun, nfev)
+
+
+def cobyla(m_nlcon, f, n, x0, m_ineq=0, Aineq=[], bineq=[], m_eq=0, Aeq=[], beq=[], xl=None, xu=None, rhobeg=1.0, rhoend=1e-3, maxfun=1000, iprint=0):
```

Comment:
To ease the interface with existing Python libraries, isn't it possible to let the user give the objective and constraint functions, and perform the amalgamation inside the Python function `cobyla`?

Reply:
yes, I see scipy separates the gradients also, so lets do it